light-client/src/supervisor.rs

```diff
@@ -0,0 +1,302 @@
+
+                            for fork in forks {
+                                match fork {
+                                    Fork::Forked(block) => {
+                                        self.report_evidence(&block);
+                                        forked.push(block.provider);
+                                    }
+                                    Fork::Faulty(block) => {
+                                        self.peers.remove_secondary(&block.provider);
+                                    }
+                                }
+                            }
+
+                            if !forked.is_empty() {
+                                // Fork detected, exiting
+                                bail!(ErrorKind::ForkDetected(forked))
+                            }
+                        }
+                        None => {
+                            // No fork detected, exiting
+                            // TODO: Send to relayer, maybe the run method does this?
+                            return Ok(light_block);
+                        }
+                    }
+                }
+                // Verification failed
+                Err(_err) => {
+                    // Swap primary, and continue with new primary, if any
+                    self.peers.swap_primary()?;
+                    continue;
+                }
+            }
+        }
+
+        bail!(ErrorKind::NoValidPeerLeft)
+    }
+
+    fn report_evidence(&mut self, _light_block: &LightBlock) {
+        ()
+    }
+
+    #[pre(self.peers.primary().is_some())]
+    fn detect_forks(
+        &mut self,
+        light_block: &LightBlock,
+        trusted_state: &LightBlock,
+    ) -> Option<Vec<Fork>> {
+        if self.peers.secondaries().is_empty() {
+            return None;
+        }
+
+        let fork_detector = ProdForkDetector::new(); // TODO: Should be injectable
+        let result =
+            fork_detector.detect_forks(light_block, &trusted_state, self.peers.secondaries());
+
+        match result {
+            ForkDetection::Detected(forks) => Some(forks),
+            ForkDetection::NotDetected => None,
+        }
+    }
+
+    pub fn handler(&mut self) -> Handler {
+        Handler::new(self.sender.clone())
+    }
+
+    // Consume the instance here but return a runtime which will allow interaction
+    // Maybe return an output channnel here?
+    pub fn run(mut self) {
+        loop {
+            let event = self.receiver.recv().unwrap();
+            match event {
+                Event::Terminate(callback) => {
+                    callback.call(());
+                    return;
+                }
+                Event::VerifyToTarget(height, callback) => {
+                    let outcome = self.verify_to_target(height);
+                    callback.call(outcome);
+                }
+                _ => {
+                    // NoOp?
+                }
+            }
+        }
+    }
+}
+
+pub struct Handler {
```

Comment:
I mentioned that in the call but also noting here: we might want to name this differently. I don't have a better suggestion yet 😬

Reply:
I think this is called the Facade pattern. Dunno how comfortable we feel being that pretentious. 🧐

Reply:
Hmm, SthSthFacade wouldn't really help I guess and it sounds like we werehiding a lot of complexity behind that facade (but the design you and @romac came up with looks actually simple, in a good way). 

Maybe as soon as we see both the supervisor and handler in action we can find a better name for the latter 🤔 
From a data pov the handler is really just a wrapper around a sending channel. Similarly the `handler` method "just" enables you to access that channel.  We can brainstorm naming in a followup issue or on a call I guess.

Reply:
Yeah I'd this is more a handle (as in thread handle) than a handler (as in event handler). So what about `Handle` or `LightClientHandle`?

Reply:
Yes, handle is less confusing 👍